### ISSUE TITLE
Does not support Jpeg2000 decompression?

### ISSUE THREAD

**Issue by cslee99** [Efferent-Health/fo-dicom.Codecs#8]
An exception occurs when decompressing Jpeg2000 DICOM.

**Exception** : Error in JPEG 2000 code stream!
at Dicom.Imaging.Codec.DicomJpeg2000NativeCodec.Decode(DicomPixelData oldPixelData, DicomPixelData newPixelData, DicomCodecParams parameters)
at Dicom.Imaging.Codec.DicomTranscoder.Decode(DicomDataset oldDataset, DicomTransferSyntax outSyntax, IDicomCodec codec, DicomCodecParams parameters)
at Dicom.Imaging.Codec.DicomTranscoder.Transcode(DicomDataset dataset)
at Dicom.Imaging.Codec.DicomTranscoder.Transcode(DicomFile file)
at DCMTool.frmMain.searchFolderFiles(Object folderList) in D:\Dev\DICOM_Decompressor\DCMDecompressor\Form1.cs:line 225

**Steps to reproduce the behavior**
original file : JPEG 2000 Image Compression (Lossless Only)
source :
var transcoder = new DicomTranscoder(file.Dataset.InternalTransferSyntax, DicomTransferSyntax.ExplicitVRLittleEndian);
file = transcoder.Transcode(file);
file.Save(saveFilePath);

build environment : Debug x64 or Release x64

**fo-dicom version and OS/platform**
windows10, etc windows

**Comment by taprm** [comment]
I have the same problem. Can I do anything to fix it?

**Comment by cslee99** [comment]
> I have the same problem. Can I do anything to fix it?

Did you solve it?

**Comment by cslee99** [comment]
> > > I have the same problem. Can I do anything to fix it?
> > 
> > 
> > Did you solve it?
> 
> Unfortunately no. But could you share a decision, if you manage to solve it?
> I tried getting the compressed pixel data as it is, and decompressing with one of cross-platform .net core libraries for image processing.
> But unfortunately no library works, cause it seems to me that compressed pixel data doesn't contain any headers which are necessary for jpeg.
> I think it's worth trying looking through the original fo-dicom sources to find out how it processes compressed pixel data

Are you unable to decompress with any compression format?
I didn't understand exactly what the problem was.
If you share the DICOM file, I will test it.

**Comment by jaime-olivares** [comment]
Guys, the error message reveals you are invoking the original fo-dicom codec, not this one. 
Are you invoking the following call at the beginning of your application, as suggested in the readme?

```sharp
Dicom.Imaging.Codec.TranscoderManager.SetImplementation(new Efferent.Native.Codec.NativeTranscoderManager());
```

**Comment by taprm** [comment]
> Guys, the error message reveals you are invoking the original fo-dicom codec, not this one.
> Are you invoking the following call at the beginning of your application, as suggested in the readme?
> 
> ```
> Dicom.Imaging.Codec.TranscoderManager.SetImplementation(new Efferent.Native.Codec.NativeTranscoderManager());
> ```

@jaime-olivares Yes, I use exactly the same as in the Readme. First of all, I initialize codes as you mentioned, and after that I try to get pixel data this way:
dicomFile.Clone(DicomTransferSyntax.DeflatedExplicitVRLittleEndian);
When this code is executed, I get an error "Error in JPEG2000 code stream!" with the following stack trace
   at Efferent.Native.Codec.DicomJpeg2000NativeCodec.Decode(DicomPixelData oldPixelData, DicomPixelData newPixelData, DicomCodecParams parameters)
   at Dicom.Imaging.Codec.DicomTranscoder.Decode(DicomDataset oldDataset, DicomTransferSyntax outSyntax, IDicomCodec codec, DicomCodecParams parameters)
   at Dicom.Imaging.Codec.DicomTranscoder.Transcode(DicomDataset dataset)
   at Dicom.Imaging.Codec.DicomTranscoder.Transcode(DicomFile file)
   at Dicom.Imaging.Codec.DicomCodecExtensions.Clone(DicomFile file, DicomTransferSyntax syntax, DicomCodecParams parameters)
   at TreatmentPlanningSystem.WebApi.Services.FooDicomProvider.GetRawData(DicomFile dicomFile) in D:\Work\TreatmentPlanningSystem\TreatmentPlanningSystem.WebApi\Services\FooDicomProvider.cs:line 156

**Comment by jaime-olivares** [comment]
we will need a sample image.
This library does support JPEG2000 compression and decompression.

**Comment by jaime-olivares** [comment]
@taprm Are you using fo-dicom desktop or fo-dicom.netcore?
I think fo-dicom.desktop (not this library) is not supporting JPEG2000.
If that's the case, you should use fo-dicom.netcore from your .net framework project.

**Comment by taprm** [comment]
> @taprm Are you using fo-dicom desktop or fo-dicom.netcore?
> I think fo-dicom.desktop (not this library) is not supporting JPEG2000.
> If that's the case, you should use fo-dicom.netcore from your .net framework project.

I'm using fo-dicom.netcore. Here is the part of *.csproj file where I try to read that DICOM file.

![image](https://user-images.githubusercontent.com/13016376/84878283-ce009280-b0b3-11ea-8a75-e1b3369c68b2.png)

**Comment by jaime-olivares** [comment]
@taprm we have some progress with this analysis.
We found that the `openjpeg` library for Windows is failing for your image. It doesn't fail in Linux though (see image below).
Also, we have tried with other JPEG2000 images in Windows and they seem to work. So we are still determining how they differ from yours.

<img width="248" alt="image" src="https://user-images.githubusercontent.com/6194083/84937775-e7004800-b0a1-11ea-9b21-fd30e67fc7de.png">

**Comment by taprm** [comment]
@jaime-olivares it looks really strange. I've tried to run the sample(link in my previous post) on Ubuntu 20 using Rider IDE and it failed too. Here is a screenshot of this:

![image](https://user-images.githubusercontent.com/13016376/84969232-778e6500-b142-11ea-91e9-4167b132fb91.png)

So it behaves the same running on both Windows and Ubuntu. Which OS are you using?

**Comment by gofal** [comment]
@jaime-olivares Would you be so kind and check this file:
[IMG006anon.zip](https://github.com/Efferent-Health/fo-dicom.Codecs/files/7141725/IMG006anon.zip)
It contains a Jpeg2k file that can be perfectly read with other viewers, but not with fo-dicom using nativecodec. The same exception "Error in JPEG 2000 code stream!" raises with our production code in docker alpine linux as well on my devleopment machine windows10. Such files arrive from time to time, where decoding jpeg2k with your nativecodec fails.

**Comment by jaime-olivares** [comment]
Hi @gofal 
We will check right now. Please give me some clues:
Is that file working correctly with any other implementation? so we can compare phase by phase.
This may take some time to debug.

**Comment by gofal** [comment]
This image does not work with fo-dicom version 4, based on the old native codec that worked only in windows, too. But the image works in all other viewers that I tried.

But it is not surprising that it also did not work with prior fo-dicom implementations. The codec was implemented by Andersson in 2015 (https://github.com/fo-dicom/fo-dicom/issues/74) and is based on OpenJpeg 1.5.3 - current release is OpenJpeg 2.4.0. So there may be quite some bugs/issues contained.
Did you update to latest Version when you migrated the codec from fo-dicom?

**Comment by jaime-olivares** [comment]
Update:
We have determined that OpenJpeg 2.4 has a small impact (but still has) over our existing codebase.
Next analysis point: Determine if 2.4 solves the exception with the provided image

**Comment by jaime-olivares** [comment]
Update:
We have determined that 2.4 performs the transcoding correctly. So, it was an issue with the previous version (1.5). 
Now we are working on using that new version in our codecs. Not a trivial task, as the methods' footprint has changed.

**Comment by jaime-olivares** [comment]
Hi @gofal,
The issue seems to be solved in version 5.0.2 (https://www.nuget.org/packages/fo-dicom.Codecs/5.0.2)
Here is the explanation:
OpenJPEG 1.5 doesn't auto-detect the sub-format of the image (either JP2 or J2K), while version 2.4 does correctly.
As OpenJPEG 2.4 has a different footprint than 1.5, we have just added a detection mechanism to 1.5 in the C++ code.
Now it is working as expected.
Please test at your end and let us know.
